### PR TITLE
Use ArrayPool

### PR DESCRIPTION
**Pull request by pranavkm**
FYI, this one's using pre-rendering. I haven't gotten a workflow to do profiling without it as yet.

Before:

image

After:
image

### CODE REVIEW

**Comment by SteveSandersonMS** [comment]
`RenderTreeBuilder` also instantiates an `ArrayBuilder<RenderTreeFrame>` (`_entries`), but it doesn't look like anything in this PR would dispose it.

We also have various unit tests that instantiate `RenderBatchBuilder` and `ComponentState`. Is it worth ensuring they also dispose them?

File: src/Components/Components/src/Rendering/Renderer.cs [hunk 1]
```diff
@@ -710,18 +710,17 @@ protected virtual void Dispose(bool disposing)
...
-                if (componentState.Component is IDisposable disposable)
+                try
+                {
+                    componentState.Dispose();
+                }
+                catch (Exception exception)
                 {
-                    try
-                    {
-                        disposable.Dispose();
```
Comment (SteveSandersonMS) [inline src/Components/Components/src/Rendering/Renderer.cs]: This is not the only place where components get disposed. This code path is what happens when the renderer itself is disposed (which disposes all remaining components in it), but we also dispose components individually when they get removed from their parent. We'll need some more logic in `ComponentState.DisposeInBatch` to ensure it disposes the entire `ComponentState` now.

**Review by SteveSandersonMS — CHANGES_REQUESTED** [review]
This looks promising! Requesting changes as per comments.

File: src/Components/Components/src/Rendering/Renderer.cs (hunk 1 above)
Comment (pranavkm) [reply src/Components/Components/src/Rendering/Renderer.cs]: Unrelated, but should Dispose not Dispose the item if `_componentWasDisposed` was set to true in `DisposeInBatch`?

**Comment by pranavkm** [comment]
Noticed a super small bit that involves an API change - could we make [ArrayRange.Clone](https://github.com/aspnet/AspNetCore/blob/master/src/Components/Components/src/RenderTree/ArrayRange.cs#L37-L42) internal? It would give us the flexibility to change it in the future to be backed by a ArrayPool.

**Comment by SteveSandersonMS** [comment]
A couple of perf ideas:

- if the arrays are pooled, we could make the initial capacity a lot bigger, since we’re no longer so concerned about triggering more frequent GC. For diffs etc, it might be better to start at size 100. Ideally, some benchmarking might indicate whether this is an improvement. 

- Instead of asking the pool to clear the arrays when they are returned, we could clear them ourselves, since we know how many items in the array are actually no empty already. Again, benchmarking would be the real judge, but I’d guess it might be faster to clear just the first 8 items in a 1000 item array than all of them.

**Comment by rynowak** [comment]
The suggestion about clearing is probably correct. The default pool just uses `Array.Clear` with the full length of the array. so it's definitely possible for us to avoid work if we manage it ourselves.

**Comment by pranavkm** [comment]
🆙 📅

**Comment by SteveSandersonMS** [comment]
Before merging this, can you check using the E2EPerf benchmarks what effect this has on the size and runtime performance of client-side Blazor? I'm hoping that the linker means that `System.Buffers` gets reduced to a small amount so doesn't bloat the app too much, and the rendering performance improves rather than degrades with this change.

Let me know if you need any pointers with doing that.

File: src/Components/Components/src/RenderTree/ArrayBuilder.cs [hunk 2]
```diff
@@ -15,26 +17,25 @@ namespace Microsoft.AspNetCore.Components.RenderTree
     /// components can be long-lived and re-render frequently, with the rendered size
     /// varying dramatically depending on the user's navigation in the app.
     /// </summary>
-    internal class ArrayBuilder<T>
+    internal class ArrayBuilder<T> : IDisposable
     {
-        private const int MinCapacity = 10;
+        // The following fields are memory mapped to the WASM client. Do not re-order or use auto-properties.
```
Comment (SteveSandersonMS) [inline src/Components/Components/src/RenderTree/ArrayBuilder.cs]:
```suggestion
        // The following fields are memory mapped to the WASM client. Do not re-order or use auto-properties, or add more fields, without also updating the corresponding TypeScript code.
```

Comment (SteveSandersonMS) [reply src/Components/Components/src/RenderTree/ArrayBuilder.cs]: Actually ignore this. Adding fields wouldn't break it (if I remember correctly).

**Comment by pranavkm** [comment]
@SteveSandersonMS here's what I get from running the E2E perf. Is that what you're looking for?

image

**Comment by SteveSandersonMS** [comment]
@pranavkm Yes, but I was hoping for both "before" and "after" numbers so we can see what effect this change has (with both sides running on the same hardware, same browser version, etc.).

Also if you can determine the total app size for StandaloneApp with and without your change then we'll know if there's any meaningful size cost to consider. We don't bother checking this for every single change, but when you're bringing in an entirely new assembly it's worth looking at.

Hopefully this won't take too long to do, but I'm happy to get involved if needed.

**Comment by pranavkm** [comment]
image

**Comment by pranavkm** [comment]
Before:

image


After:

image

**Comment by pranavkm** [comment]
Firefox:

Before:

image

After:
image

File: src/Components/Components/src/RenderTree/ArrayBuilder.cs [hunk 3]
```diff
@@ -95,34 +96,51 @@ internal int Append(T[] source, int startIndex, int length)
         /// <param name="value">The value.</param>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public void Overwrite(int index, in T value)
-            => _items[index] = value;
+        {
+            if (index > _itemsInUse)
```
Comment (SteveSandersonMS) [inline src/Components/Components/src/RenderTree/ArrayBuilder.cs]: Shouldn't this be `if (index >= _itemsInUse)`? If `_itemsInUse == 0`, then this should throw even if `index` is `0`.

Thanks for adding the new unit tests for `ArrayBuilder`. Looks like we could use one to cover the boundary case for `Overwrite` when it should throw.

File: src/Components/Components/src/Rendering/ComponentState.cs [hunk 4]
```diff
@@ -91,6 +91,8 @@ public void DisposeInBatch(RenderBatchBuilder batchBuilder)
             {
                 RemoveCascadingParameterSubscriptions();
             }
+
+            DisposeBuffers();
```
Comment (SteveSandersonMS) [inline src/Components/Components/src/Rendering/ComponentState.cs]: Any reason you can't merge `Dispose` and `DisposeBuffers` and just call `Dispose` from here, letting it take care of `component.Dispose()`? It would be nice if there was only one place where we call dispose on the component itself.

File: src/Components/Components/src/Rendering/Renderer.cs (hunk 1 above)
Comment (SteveSandersonMS) [reply src/Components/Components/src/Rendering/Renderer.cs]: We definitely shouldn't double-dispose components. My guess is that we wouldn't anyway because if `DisposeInBatch` was called, then we'd have removed that `ComponentState` from the renderer at the same time, at least in normal cases.

But I'm not against being clearer about this in the code by having `ComponentState` track whether or not it was already disposed so it definitely doesn't double-do it.

**Review by SteveSandersonMS — APPROVED** [review]
Excellent!

I know it's tricky when making changes to such low-level things and changing memory management concerns, because there are so many existing usages that have to be considered to figure out whether something would now leak or degrade perf. But this certainly looks good, and it's great to see the new unit tests for `ArrayBuilder`.

Thanks for posting the E2E perf numbers. Since it appears that nothing is meaningfully regressed I think we're all clear there. We'd expect to see perf gains from this on bigger client-side apps that would otherwise stress the GC more.
